docs/CppInteroperability/SwiftToCppInteroperabilityRoadmap.md

```diff
@@ -0,0 +1,125 @@
+
+## Goals
+
+Swift and C++ are both feature rich programming languages. While they both have a lot of similar features, Swift has some features which C++ can not represent natively. For instance, Swift supports overloading based on argument labels, but C++ does not. A gap like this can be bridged in an idiomatic manner between the two languages, as the C++ name of the exposed Swift function can be modified to reflect the Swift argument labels. However, some Swift features like result builders can not be bridged in an idiomatic manner to C++. Therefore, we can assume that it is impractical for Swift to expose all Swift APIs to C++. Additionally, some Swift API patterns, like conforming Swift types to specific protocols, require advanced compiler features that are impractical to support in a generated header. Thus, we can assume that it is also impractical for Swift to bridge all Swift API patterns into C++.
+
+For the reasons outlined above, it is not a goal to expose every Swift API and language construct to C++, and it is also not a goal to make every Swift library accessible to C++. There will certainly be libraries that will lend themselves well to an ergonomic and an idiomatic C++ interface, however, there will always be some libraries that rely on language features or API patterns that can’t be bridged to C++ in a reasonable manner. 
+
+It is a goal to allow mixed Swift and C++ projects to call their own Swift APIs that lend themselves well to interoperability from C++, and do it in a safe, performant and ergonomic manner. It is also a goal to allow these exposed Swift APIs to use Swift types from other Swift libraries (including the Swift standard library), even if the other Swift library wasn’t developed with C++ interoperability in mind. 
+
+When Swift APIs get exposed to C++, the Swift compiler and the language specification should follow several key principles that are presented below.
+
+### Safety
+
+Safety is a top priority for the Swift programming language. Swift code expects its caller to adhere to Swift’s type rules and Swift’s memory model, regardless of whether it’s called from Swift or C++. Thus, the C++ code that calls Swift should properly enforce Swift’s expected language contracts. Such enforcement is needed to minimize the risk of Swift behaving in an unexpected manner at run-time when it’s being called from C++, as such behavior could lead to buggy program behavior or crashes that a Swift programmer would not expect to see in their Swift code. This kind of enforcement does not prevent all possible issues though. For instance, a bug in user’s C++ code could accidentally overwrite a Swift object stored on the heap, which could cause unexpected behavior or crashes in Swift code. To catch bugs like this one, the user should use other program analysis tools such as address sanitizer.
```

Comment:
You talk about Swift's safety model here, but it might be good to call out explicitly: even though a C++ programmer is calling Swift APIs, C++'s safety model will not change. (I guess the example at the end sort of does that.) And you should probably call out what "C++'s safety model" means in this case: C++ often makes safety the user's responsibility, so users will have to be aware of Swift's safety model and will be responsible for enforcing Swift's model on the C++ side.

Reply:
Updated this paragraph to something like this to incorporate your feedback:


Safety is a top priority for the Swift programming language. Swift code expects its caller to adhere to Swift’s type rules and Swift’s memory model, regardless of whether it’s called from Swift or C++. Thus, the C++ code that calls Swift should properly enforce Swift’s expected language contracts. The enforcement should be done automatically in the generated header. Such enforcement is needed to minimize the risk of Swift behaving in an unexpected manner at run-time when it’s being called from C++, as such behavior could lead to buggy program behavior or crashes that a Swift programmer would not expect to see in their Swift code. This kind of enforcement does not prevent all possible issues though, as it does not change C++'s safety model. C++ is unsafe by default, and the user is able to use regular C++ pointers to write to memory as if they were using Swift's `UnsafeMutablePointer` type. This means that bugs in C++ code can easily lead to violations of Swift's expected invariants. For instance, a bug in user’s C++ code could accidentally overwrite a Swift object stored on the heap, which could cause unexpected behavior or crashes in Swift code. The user is expected to obey Swift's memory model and type rules when calling into Swift APIs from C++, and thus they bear the ultimate responsibility for avoiding bugs like this one. The user can use certain program analysis tools, such as Address Sanitizer, to help them catch bugs that violate Swift's memory model rules.